### ISSUE TITLE
Resolve JavaCore corresponding parameter names

### ISSUE THREAD

**Issue by TerryvanWalen** [javaparser/javaparser#1624]
In my application I would like to retrieve the corresponding parameter names to all MethodCallExpr arguments.

Using the following code this does not work well for correspondingMethods in the JAVA CORE packages (Scanning GoogleGuava source: https://github.com/google/guava).

```java
CombinedTypeSolver combinedSolver = new CombinedTypeSolver();

TypeSolver reflectionTypeSolver = new ReflectionTypeSolver(false);
combinedSolver.add(reflectionTypeSolver);

File rootDirectory;
for (SourceRoot s : projectRoot.getSourceRoots()) {
     rootDirectory = s.getRoot().toFile();
     TypeSolver javaParserTypeSolver = new JavaParserTypeSolver(rootDirectory);
         combinedSolver.add(javaParserTypeSolver);
}
```

However I already have a simple solution that works quite well, the problem with this however is that it is slow.

```java
CombinedTypeSolver combinedSolver = new CombinedTypeSolver();

TypeSolver javaCoreNameSolver = new JavaParserTypeSolver("C:\\Users\\****\\IdeaProjects\\JAVACORE");
combinedSolver.add(javaCoreNameSolver);

File rootDirectory;
for (SourceRoot s : projectRoot.getSourceRoots()) {
      rootDirectory = s.getRoot().toFile();
      combinedSolver.add(new JavaParserTypeSolver(rootDirectory));
}
```

(The C:\\Users\\****\\IdeaProjects\\JAVACORE directory contains all the files from the src.zip of the JDK directory)

I was wondering if anybody here knows a better way of doing this?

PS.
I also collected all packages that could not solve the name of the parameter (if the corresponding parameter in question was resolved).

Using:
```java
ResolvedParameterDeclaration parameter = correspondingDeclaration.getParam(i);

someForLoop i { 
    ResolvedParameterDeclaration parameter = correspondingDeclaration.getParam(i);

    try {
        parameter.getName();
    } catch (UnsupportedOperationException e) {
        packageCollector.getInstance().increase(packageName);
        break;
    }
}
```

java.util.logging=60,
javax.crypto=12, 
java.util=2320, 
java.util.stream=46, 
com.google.common.reflect=62, 
java.nio=72,
com.google.common.util.concurrent=312, 
java.nio.channels=12, 
com.google.common.net=8,
java.util.zip=4, 
java.nio.file=6, 
java.nio.charset=12, 
java.net=4, 
java.util.regex=18, 
com.google.common.eventbus=14,
java.math=68, 
com.google.common.collect=2682, 
sun.misc=22, 
java.lang.reflect=54, 
java.security=8, 
com.google.common.graph=462, 
com.google.common.base=3676, 
com.google.common.hash=142, 
java.util.concurrent=304, 
com.google.common.escape=18, 
java.util.concurrent.atomic=200, 
com.google.common.math=182, 
java.util.concurrent.locks=32, 
com.google.common.primitives=146, 
com.google.common.cache=118, 
java.lang=1154, 
java.io=160, 
java.util.function=154, 
com.google.common.io=142

**Comment by matozoid** [comment]
@ftomassetti - any chance you could answer this?

**Comment by ftomassetti** [comment]
I am not sure I understand it: it seems that initially the solving would not work but simply because he was using the wrong typesolver. Then it started working but he was unhappy with the performance. Am I understand it correctly?

Then there are some things that are unsolved, I suspect because the typesolver does not include the dependencies (like the com.google stuff).

**Comment by TerryvanWalen** [comment]
Hi @ftomassetti,

Thanks for looking into it.

To clarify a bit, the main problem is the ReflectionTypeSolver. Please correct me if I am wrong but I expected the ReflectionTypeSolver to solve all method calls targetting methods of JAVA CORE like methods in the java.util.* and java.io.* packages. This indeed happens (resolving the method declarations), however the parameter belonging to these solved method declarations do not possess a name.

So the problem is that the formal parameters of a resolved method declaration (in java.* packages) do not contain the names of that formal parameter if solved with a ReflectionTypeSolver.

So what I did is to include the source zip file included in the JDK (with all base java source files) and load it in as a project in a JavaParserTypeSolver but this is not an optimal solution and doesn't work together with the ReflectionTypeSolver.

Hope this clarifies what I mean a bit.

**Comment by ftomassetti** [comment]
Right! Yes, I miss the point entirely. Well, I think that parameter name can be _optionally_ included in compiled code. Maybe when they are not present we can come up with default name. @matozoid what do you think? Should we use default name or just specify in the documentation that parameter names could potentially be not present?

**Comment by matozoid** [comment]
One example of a library storing parameter names is javaparser-core, as can be seen here: https://github.com/javaparser/javaparser/blob/master/javaparser-core/pom.xml#L134

This is not done by default, and I think it is rare for people to turn this on, so not having parameter names is the default. It can be seen in various libraries that use annotations on parameters and are forced to repeat the name in an annotation (because it was lost during compilation!)

Do document this!

**Comment by ftomassetti** [comment]
We have also to check if ASM actually gave us access to the parameter name

**Comment by ftomassetti** [comment]
It turns out that ASM comes up with default names, see tests in PR #1664

**Comment by TerryvanWalen** [comment]
Looks perfect. Should I close the issue?

**Comment by ftomassetti** [comment]
It should be closed automatically when we merge the PR

**Comment by matozoid** [comment]
It gets automatically closed when you put the magic text in the PR description: "Fixes #1234"
